Wireframe

ryandthorn opened the pull request:
Looking to create todos that disappear when a timer runs out. Also need to add tabs for active/completed todos.

ryandthorn committed:
Add Material-UI core

ryandthorn committed:
Move App to src/components

ryandthorn committed:
Create TodoInput

ryandthorn committed:
Update import

ryandthorn committed:
Add Material icons

ryandthorn committed:
Create TodoList

ryandthorn committed:
Add adding todos; lift todos state to App

ryandthorn committed:
Add delete todo

ryandthorn committed:
Rename TodoInput -> TodoForm

ryandthorn committed:
todo => {text, isCompleted}

ryandthorn committed:
Replace HTML els with MUI containers

grantglidewell commented:
Im going to review the code as well, but this is a case that you should handle:
<img width="602" alt="Screen Shot 2019-09-12 at 8 46 34 PM" src="https://user-images.githubusercontent.com/26661451/64836401-94d44880-d59e-11e9-8c0b-27d9c21bbb10.png">
You are able to add empty todos.
Also might be nice to have some padding on top here.

grantglidewell commented on src/components/App.js:
```diff
@@ -0,0 +1,19 @@
+
+  return (
+    <Container className="App" maxWidth="sm">
+      <TodoForm todos={todos} setTodos={setTodos}></TodoForm>
```
This component doesnt need to know about the todos in the list.

grantglidewell commented on src/components/App.js:
```diff
@@ -0,0 +1,19 @@
+  return (
+    <Container className="App" maxWidth="sm">
+      <TodoForm todos={todos} setTodos={setTodos}></TodoForm>
+      <TodoList todos={todos} setTodos={setTodos}></TodoList>
```
This component is only responsible for updating existing todos. It might make sense to have functions that are specific to completing and deleting and pass those in, it's a more explicit API.

grantglidewell commented on src/components/TodoForm.js:
```diff
@@ -0,0 +1,44 @@
+import InputLabel from "@material-ui/core/InputLabel";
+
+const TodoForm = props => {
+  const [todoText, setTodoText] = React.useState("");
```
This may not be necessary, unless the input only works as a controlled component. I believe you could accomplish this with an uncontrolled component. If there is a specific reason youre doing this let me know, there could be valid reasons.

grantglidewell commented on src/components/TodoForm.js:
```diff
@@ -0,0 +1,44 @@
+    <Grid
+      container
+      component="form"
+      onSubmit={e => handleSubmit(e)}
```
nit pick: e is not explicit, is it error? is it the event as it is here? short variable names are for golang and rust, not javascript. Dont be affraid to be explicit. I would call this `evt`

grantglidewell commented on src/components/TodoList.js:
```diff
@@ -0,0 +1,52 @@
+import DeleteIcon from "@material-ui/icons/Delete";
+import DoneIcon from "@material-ui/icons/Done";
+
+const TodoList = props => {
```
Its a style choice, but destructuring props is a common standard in most modern React apps, makes it very clear what is being used

grantglidewell commented on src/components/TodoList.js:
```diff
@@ -0,0 +1,52 @@
+        const labelId = `todo-label-${index}`;
+        return (
+          <ListItem key={index} role={undefined} dense>
+            <ListItemText id={labelId} primary={todo.text} />
```
this _may_ become problematic if you ever render multiple lists this way. the labeId has the potential to have a matching ID in another part of the application. it's be to make sure they are unique. One way might be combining the todo text with the index. But using the index is generally a very last resort.

grantglidewell reviewed (commented):
Overall this is a great start!

ryandthorn committed:
Create useTodo API; persist with localStorage

ryandthorn committed:
Using Date.now to generate ID & expiry

- IDs are now Date.now() millisecond strings
- Expiry is current ms + duration in ms
- Added a hook to set a timer to check for expiry

ryandthorn committed:
Remove placeholder tests

ryandthorn committed:
Create SetTodoTime; can now input todo duration

ryandthorn committed:
Add tabs

- add property todo.isExpired
- filter active, complete, and missed todos with tabs
- add MUI's rec'd display meta info

ryandthorn committed:
Add styling; todos show time left

ryandthorn committed:
Correct how time left is displayed

grantglidewell commented on src/components/SetTodoTime.js:
```diff
@@ -0,0 +1,23 @@
+    <Grid container item justify="space-evenly">
+      <Grid item>
+        <InputLabel htmlFor="hours">Hours</InputLabel>
+        <Input id="hours" className="Input__time" type="number" />
```
Is this className defined somewhere?

grantglidewell commented on src/components/TodoList.js:
```diff
@@ -18,14 +18,38 @@ const TodoList = ({ todos, deleteTodo, completeTodo }) => {
+      break;
+    case 2:
+      filteredTodos = todos.filter(
+        todo => todo.isCompleted === false && todo.isExpired === true
```
this is interesting, lets refactor it.

ryandthorn replied:
I replaced the index numbers with explicit strings for active, completed, and missed todos.

grantglidewell commented on src/components/TodoList.js:
```diff
@@ -18,14 +18,38 @@ const TodoList = ({ todos, deleteTodo, completeTodo }) => {
+  }
+
+  const renderCompleteButton = todo =>
+    todo.isCompleted === true || todo.isExpired === true ? false : true;
```
this is quite hard to reason about, lets refactor it

ryandthorn replied:
Thanks for your advice with and/or chaining!

grantglidewell commented on src/components/TodoList.js:
```diff
@@ -40,9 +53,21 @@ const TodoList = ({ todos, deleteTodo, completeTodo, activeTab }) => {
+      ? `${Math.floor(
+          (todo.expiration - Date.now()) / 60000
+        )} minutes remaining`
+      : null;
```
also overly terse here.

ryandthorn committed:
Replace numbered tab indices w/ explicit strs

ryandthorn committed:
Extract funcs and add comments for readability
